h2o-core/src/main/java/water/persist/PersistManager.java

```diff
@@ -261,6 +264,15 @@ private static boolean httpUrlExists(String URLName){
     return I[Value.NFS].calcTypeaheadMatches(filter, limit);
   }
 
+  private Key frameFromHTTP(String path) throws IOException {
+    java.net.URL url = new URL(path);
+    Key destination_key = Key.make(path);
+    java.io.InputStream is = url.openStream();
+    UploadFileVec.ReadPutStats stats = new UploadFileVec.ReadPutStats();
+    UploadFileVec.readPut(destination_key, is, stats);
```

Comment:
This will actually read the data in memory. This is not consistent with the behavior of `anyURIToKey`

It is probably the easiest solution to solve this issue. It might be okay, but we need to be aware of the consequences.

Reply:
We could use HTTP/1.1 Range header and implement fully distributed HTTP ingest (https://en.wikipedia.org/wiki/Byte_serving)

This wouldn't be in the scope of this PR and it wouldn't work for everything.

Reply:
@michalkurka Good point, I didn't know about Byte serving. I was just following already implemented path in the code. But I can have a look for sure on the improvement of this solution, but maybe in different PR?

Reply:
I probably don't understand this point properly. We do not have a concept of lazy data import in H2O, do we? In this particular case we load the data when the user asks for it

Reply:
Normal parse works lazily: you first create the metadata in memory (FileVec, HdfsVec,  ...) and when parse is executed the data are loaded in memory.

Your PR changes the semantics of anyUriToKey - for HTTP it also loads the data. I think a good solution here is to handle HTTP as a special case without modifying anyUriToKey.

I would fix it on the caller side (where we know the purpose is to parse the data, we do not know that in anyUriToKey):

  public static Frame parseFrame(Key okey, ParseSetup parseSetup, URI ...uris) throws IOException {
    if (uris == null || uris.length == 0) {
      throw new IllegalArgumentException("List of uris is empty!");
    }
    if(okey == null) okey = Key.make(uris[0].toString());
    Key[] inKeys = new Key[uris.length];
    for (int i=0; i<uris.length; i++)  inKeys[i] = H2O.getPM().anyURIToKey(uris[i]);
    // Return result
    return parseSetup != null ? ParseDataset.parse(okey, inKeys, true, ParseSetup.guessSetup(inKeys, parseSetup))
                              : ParseDataset.parse(okey, inKeys);
  }

Does that make sense?

Reply:
I think the special handling in FrameUtils makes sense for now. Eventually, we could implement "PersistHttp" that would either do a distributed ingest (with range) or just simple ingest depending on if the server supports it.

WDYT?

Reply:
Sorry for delay and thanks for great explanation! Yup, agree, I'm working on a change, still WIP, need to test it properly. Will let you know( I need to first finish different task on SW, but that should be done today)